### PR TITLE
ISAAC week3 HW

### PR DESCRIPTION
**Pull request by gisaac85**

### CODE REVIEW

File: week3/homework/package.json [hunk 1]
```diff
@@ -13,5 +13,11 @@
     "eslint-plugin-node": "^6.0.0",
     "eslint-plugin-promise": "^3.5.0",
     "eslint-plugin-standard": "^3.0.1"
+  },
+  "dependencies": {
+    "express": "^4.16.3",
+    "fs": "0.0.1-security",
```
Comment (GeorgeSapkin) [inline week3/homework/package.json]: `fs` is a built-in module you don't need to install it using `npm`.

File: week2/homework/src/index.js [hunk 2]
```diff
@@ -1,3 +1,3 @@
 'use strict';
 
-// TODO: Write the homework code in this file
+// TODO: Write the homework code in this file
```
Comment (GeorgeSapkin) [inline week2/homework/src/index.js]: Why did you change this file from week 2 here?

**Comment by GeorgeSapkin** [comment]
You can push changes to the same branch and they will appear in this PR.

**Comment by gisaac85** [comment]
But is my PR is in good acceptable state??

File: week3/homework/src/index.js [hunk 3]
```diff
@@ -1,3 +1,192 @@
...
+//READ ALL TODOs
+app.get('/todos', async (req, res) => {
+  const todos = await readTodos();
+  res.json(todos);
+  res.end();
+});
+
+//READ TODO BY ID
+app.get('/todos/:id', async (req, res) => {
+  try {
```
Comment (alxbrg) [inline week3/homework/src/index.js]: These `try` and `catch` blocks are a bit weird to me. They are meant to catch run-time errors. We should not get any of those here if our code checks out.

File: week3/homework/src/index.js [hunk 4]
```diff
@@ -1,3 +1,192 @@
...
+    const id = req.params.id;
+    const todos = await readTodos();
+    const todo = todos.find(x => x.id === id);
+    if (typeof todo === 'undefined') {
+      res.json({
+        'Error Message': 'NO such ID !!'
+      });
+    }
+    res.json(todo);
```
Comment (alxbrg) [inline week3/homework/src/index.js]: Since you already returned a response above if the `todo` was not found, you should not call `res.json` again. This part of the code should be unreachable if we don't need/want it to be executed. What you could do, is `return` on line 64. Does that make sense?

File: week3/homework/src/index.js [hunk 5]
```diff
@@ -1,3 +1,192 @@
...
+    res.json({
+      'Error Name': err.name,
+      'Error Message': err.message
+    });
+  }
+
+});
+
+//DELETE TODO BY ID
+app.delete('/todos/:id', async (req, res) => {
```
Comment (alxbrg) [inline week3/homework/src/index.js]: Comments above also apply here.

File: week3/homework/src/index.js [hunk 6]
```diff
@@ -1,3 +1,192 @@
...
+  }
+
+});
+
+//markAsNotDone TODO
+app.delete('/todos/:id/done', async (req, res) => {
+  try {
+    const id = req.params.id;
+    const todos = await readTodos();
+    const newTodos = todos.find(todos => todos.id === id);
```
Comment (alxbrg) [inline week3/homework/src/index.js]: Besides the changes you need to make, I see the same lines of code repeating. Maybe you could have some functions to help you do these actions on the todos. For instance a `findTodo` function which would accept the `request` as a param and return either throw an error (which you could `catch` 😉 ) or the `todo` that was found.

File: week3/homework/src/index.js (hunk 3 above)
Comment (gisaac85) [reply week3/homework/src/index.js]: done!

File: week3/homework/src/index.js [hunk 7]
```diff
@@ -1,3 +1,192 @@
...
+//CREATE TODO
+app.post('/todos', async (req, res) => {
+  const newTodo = req.body;
+  newTodo.id = uuid();
+  newTodo.done = false;
+  const todos = await readTodos();
+  todos.push(newTodo);
+  await writeTodos(todos);
+  res.json(todos);
+  res.end();
```
[the remark this answers is not in the document]
Comment (gisaac85) [reply week3/homework/src/index.js]: done!

File: week3/homework/src/index.js (hunk 4 above)
Comment (gisaac85) [reply week3/homework/src/index.js]: can i use res.send on line 64 when ID is not exist !!?

File: week3/homework/src/index.js [hunk 8]
```diff
@@ -1,3 +1,107 @@
...
+
+//DELETE TODO BY ID
+app.delete('/todos/:id', async (req, res) => {
+
+  const [, , ,
+    todo,
+    todos
+  ] = await findTodo(req);
+  if (typeof todo === 'undefined') {
+    return new Error(res.send('NO SUCH ID'));
```
Comment (alxbrg) [inline week3/homework/src/index.js]: See previous comment.

File: week3/homework/src/index.js [hunk 9]
```diff
@@ -1,3 +1,107 @@
...
+    id,
+    bodyReq,
+    done,
+    todo,
+    todos
+  ] = await findTodo(req);
+  const newTodo = bodyReq;
+  newTodo.id = id;
+  if (typeof todo === 'undefined') {
+    return new Error(res.send('NO SUCH ID'));
```
Comment (alxbrg) [inline week3/homework/src/index.js]: See previous comment.

File: week3/homework/src/functions.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+  return writeFile(TODO_FILE, JSON.stringify(todos));
+}
+
+//Find Todo
+async function findTodo(req) {
+  const id = req.params.id;
+  const todos = await readTodos();
+  const todo = todos.find(todos => todos.id === id);
+  const done = req.body.done;
+  return await [
```
Comment (alxbrg) [inline week3/homework/src/functions.js]: `return await` does not really make sense. You should simple `return`. You will have to `await` the resolution of the function when you call it.

File: week3/homework/src/functions.js [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+function writeTodos(todos) {
+  return writeFile(TODO_FILE, JSON.stringify(todos));
+}
+
+//Find Todo
+async function findTodo(req) {
+  const id = req.params.id;
+  const todos = await readTodos();
+  const todo = todos.find(todos => todos.id === id);
+  const done = req.body.done;
```
Comment (alxbrg) [inline week3/homework/src/functions.js]: Not sure what's the point of this. Can you explain?

File: week3/homework/src/functions.js [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+//MarkTodo Function
+async function markTodo(req, res, doneFlag) {
+  const [, ,
+    done,
+    todo,
+    todos
+  ] = await findTodo(req);
+  if (typeof todo === 'undefined') {
+    return new Error(res.send('NO SUCH ID'));
```
Comment (alxbrg) [inline week3/homework/src/functions.js]: Returning a `new Error` here does not make sense to me. Simply return `res.send('NO SUCH ID')`. Makes sense? Also: you should return a json for the errors as well, like so: `{ error: 'NO SUCH ID' }`. It will help us use that error when using this in the browser when we code our front-end application.

File: week3/homework/src/index.js [hunk 13]
```diff
@@ -1,3 +1,107 @@
...
+  res.json(todos);
+
+});
+
+//READ TODO BY ID
+app.get('/todos/:id', async (req, res) => {
+
+  const [, , , todo, ] = await findTodo(req);
+  if (typeof todo === 'undefined') {
+    return new Error(res.send('NO SUCH ID'));
```
Comment (alxbrg) [inline week3/homework/src/index.js]: See previous comment.

File: week3/homework/src/functions.js [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    done,
+    todo,
+    todos
+  ];
+}
+
+//MarkTodo Function
+async function markTodo(req, res, doneFlag) {
+  const [, ,
+    done,
```
Comment (alxbrg) [inline week3/homework/src/functions.js]: You're not using this variable anywhere.

File: week3/homework/src/functions.js [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+  const todos = await readTodos();
+  const todo = todos.find(todos => todos.id === id);
+  const done = req.body.done;
+  return await [
+    id,
+    req.body,
+    done,
+    todo,
+    todos
+  ];
```
Comment (alxbrg) [inline week3/homework/src/functions.js]: Not sure why you return so many variables 😊 I would have just returned `todo` and `todos`.

File: week3/homework/src/functions.js (hunk 14 above)
Comment (gisaac85) [reply week3/homework/src/functions.js]: when i try to delete ( , , done , ) i get error !!

File: week3/homework/src/functions.js (hunk 15 above)
Comment (gisaac85) [reply week3/homework/src/functions.js]: because i need them in markTodo and updatTodo functions !!!

**Comment by gisaac85** [comment]
can you please see my last changing !!

File: week3/homework/src/functions.js (hunk 14 above)
Comment (alxbrg) [reply week3/homework/src/functions.js]: Your `findTodo` should only return what you need and are going to use. So `findTodo` should only return a `todo` and the `todos` (I don't think you are using `id`, `req.body` and `done` anywhere).

**Comment by gisaac85** [comment]
Hi @alxbrg  
thanks for your constructive feedback..
Yes I see that returning [todo,todos] from function findTodo() are enough to do all tasks 💯 
I hope now you accept my PR 👍
